pkg/kf/algorithms/algorithms.go

```diff
@@ -0,0 +1,202 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package algorithms
+
+import (
+	"sort"
+)
+
+// Interface is the interface used by the algorithms.
+type Interface interface {
+	sort.Interface
+
+	// Set stores the value located at b[j] to a[i].
+	Set(i int, a Interface, j int, b Interface)
+
+	// Slice returns a slice (e.g., s[i:j]) of the object.
+	Slice(i, j int) Interface
+
+	// Append returns i and a appended to each other (e.g., append(i, a...)).
+	Append(a Interface) Interface
+
+	// Clone returns a clone of the object.
+	Clone() Interface
+}
+
+// Strings implements Interface for a slice of strings.
+type Strings []string
+
+// Len implements Interface.
+func (s Strings) Len() int {
+	return len(s)
+}
+
+// Swap implements Interface.
+func (s Strings) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+
+// Less implements Interface.
+func (s Strings) Less(i, j int) bool {
+	return s[i] < s[j]
+}
+
+// Set implements Interface.
+func (s Strings) Set(i int, a Interface, j int, b Interface) {
+	a.(Strings)[i] = b.(Strings)[j]
+}
+
+// Clone implements Interface.
+func (s Strings) Clone() Interface {
+	return append(Strings{}, s...)
+}
+
+// Append implements Interface.
+func (s Strings) Append(a Interface) Interface {
+	return append(s, a.(Strings)...)
+}
+
+// Slice implements Interface.
+func (s Strings) Slice(i, j int) Interface {
+	return s[i:j]
+}
+
+// Ints implements Interface for a slice of ints.
+type Ints []int
+
+// Len implements Interface.
+func (s Ints) Len() int {
+	return len(s)
+}
+
+// Swap implements Interface.
+func (s Ints) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+
+// Less implements Interface.
+func (s Ints) Less(i, j int) bool {
+	return s[i] < s[j]
+}
+
+// Set implements Interface.
+func (s Ints) Set(i int, a Interface, j int, b Interface) {
+	a.(Ints)[i] = b.(Ints)[j]
+}
+
+// Clone implements Interface.
+func (s Ints) Clone() Interface {
+	return append(Ints{}, s...)
+}
+
+// Append implements Interface.
+func (s Ints) Append(a Interface) Interface {
+	return append(s, a.(Ints)...)
+}
+
+// Slice implements Interface.
+func (s Ints) Slice(i, j int) Interface {
+	return s[i:j]
+}
+
+// Dedupe removes any duplicates from the given collection. This DOES alter
+// the input. It outputs the lengh of the new collection. Therefore, if the
+// interface is wrapping a slice, then the slice should be truncated via the
+// result (e.g., slice[:returnValue]).
+func Dedupe(s Interface) int {
+	sort.Sort(s)
```

Comment:
This function would be more powerful if the dedup was guaranteed to happen in some specific order e.g. the first/last will always be chosen. StableSort would help with that